### PR TITLE
Smoltcp bitide instance

### PR DESCRIPTION
**Pull request by lmbollen**
A bittide instance containing Vex risc and all components necessary for ethernet communication

### CODE REVIEW

File: bittide/src/Bittide/Ethernet/Mac.hs [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+    miiSel txClkEna rxClkEna txAxiData txAxiKeep txAxiValid txAxiLast txAxiUser
+    gmiiRxData' gmiiRxValid' gmiiRxError' rxAxiReady
+
+  ethStatus = makeEthStatus <$> txFifoUnderflow <*> txFifoOverflow <*> txFifoBadFrame
+    <*> txFifoGoodFrame <*> rxBadFrame <*> rxBadFcs <*> rxFifoOverflow <*> rxFifoBadFrame
+    <*> rxFifoGoodFrame
+
+  rxAxiM2S = makeAxi <$> rxAxiValid <*> rxAxiData <*> rxAxiKeep <*> rxAxiUser <*> rxAxiLast
+
+-- | Utility function to create an Axi4StreamM2S from it's components.
```
Comment (martijnbastiaan) [inline bittide/src/Bittide/Ethernet/Mac.hs:188]:
```suggestion
-- | Utility function to create an Axi4StreamM2S from its components.
```

**Review by martijnbastiaan — CHANGES_REQUESTED** [review]
Could you please remove the unused code / move it to another draft PR?

File: bittide-shake/src/Clash/Shake/Vivado.hs [hunk 2]
```diff
@@ -68,6 +69,8 @@ meetsTiming path = andM
   , fmap not $ inFile "Timing constraints are not met" path
   ]
 
+type GlobPattern = String
```
Comment (martijnbastiaan) [inline bittide-shake/src/Clash/Shake/Vivado.hs:72]: Perhaps:

```haskell
type TclGlobPattern = String
```

And make sure to point out https://www.tcl.tk/man/tcl8.6/TclCmd/glob.htm

File: bittide-instances/src/Bittide/Instances/Pnr/Ethernet.hs [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        [PortName "TX_P", PortName "TX_N"]
+      ]
+  }) #-}
+
+
+-- TODO: Add more constraints?
+-- | Convert a reset signal from one domain to another domain and insert a register.
+-- This can be used to move a reset signal from a domain with a synchronous reset to
+-- a domain with an asynchronous reset.
+unsafeResetDesynchronizer ::
```
Comment (martijnbastiaan) [inline bittide-instances/src/Bittide/Instances/Pnr/Ethernet.hs:181]: It's still synchronizing, just to other specifications. I also think it's not quite correct; you want to sample in the target domain with multiple ffs _or_ specify that users should be very careful and make sure that these domains have the same clock. Unless there is a good reason to care about that single flipflop, just use `xpmResetSynchronizer`.

**Review by martijnbastiaan — APPROVED** [review]
In general LGTM, could you look at my comments?

File: bittide-instances/src/Bittide/Instances/Pnr/Ethernet.hs (hunk 3 above)
Comment (lmbollen) [reply bittide-instances/src/Bittide/Instances/Pnr/Ethernet.hs:181]: Actually we don't want any logic or state in the "target" domain.
We use this for the asynchronous reset port on of `gmiiSgmiiBridge` which, according to https://docs.amd.com/v/u/en-US/gig_eth_pcs_pma_ds264 : 
| Signal | Direction | Clock Domain | Description                                          |
|--------|-----------|--------------|------------------------------------------------------|
| reset  | Input     | n/a          | Asynchronous reset for the entire core. Active High. |

The signal is not supposed to be synchronous, but we can't express that in Clash....

Comment (martijnbastiaan) [reply bittide-instances/src/Bittide/Instances/Pnr/Ethernet.hs:181]: `delayReset` (filter glitches) + `unsafeResetSynchronizer` (change domain)

Comment (martijnbastiaan) [reply bittide-instances/src/Bittide/Instances/Pnr/Ethernet.hs:181]: `unsafeResetSynchronizer = unsafeFromBla . unsafeSynchronizer . unsafeToBla`

**Pull request merged**